packages/@ember/-internals/glimmer/lib/helpers/action.ts

```diff
@@ -349,55 +353,51 @@ function makeArgsProcessor(valuePathRef: Reference | false, actionArgsRef: Refer
 
 function makeDynamicClosureAction(
   context: object,
-  targetRef: Reference<MaybeActionHandler>,
-  actionRef: Reference<string | AnyFn>,
+  targetRef: Reference<unknown>,
+  actionRef: Reference<unknown>,
   processArgs: (args: unknown[]) => unknown[],
   debugKey: string
 ) {
+  const action = valueForRef(actionRef);
+
   // We don't allow undefined/null values, so this creates a throw-away action to trigger the assertions
   if (DEBUG) {
-    makeClosureAction(
-      context,
-      valueForRef(targetRef),
-      valueForRef(actionRef),
-      processArgs,
-      debugKey
-    );
+    makeClosureAction(context, valueForRef(targetRef), action, processArgs, debugKey);
   }
 
   return (...args: any[]) => {
     return makeClosureAction(
       context,
       valueForRef(targetRef),
-      valueForRef(actionRef),
+      action,
       processArgs,
       debugKey
     )(...args);
   };
 }
 
 interface MaybeActionHandler {
-  actions?: Record<string, AnyFn>;
+  actions?: Record<string, Function>;
 }
 
 function makeClosureAction(
   context: object,
-  target: MaybeActionHandler,
-  action: string | AnyFn,
+  target: unknown,
+  action: unknown | null | undefined | string | Function,
   processArgs: (args: unknown[]) => unknown[],
   debugKey: string
 ) {
   let self: object;
-  let fn: AnyFn;
+  let fn: Function;
 
   assert(
     `Action passed is null or undefined in (action) from ${target}.`,
     action !== undefined && action !== null
   );
 
   if (typeof action === 'string') {
-    self = target;
-    fn = (target.actions && target.actions[action as string])!;
+    self = target as object;
+    fn = (target as { actions: Record<string, Function> })?.actions?.[action as string] as Function;
```

Comment:
```suggestion
    const value = (target as { actions?: Record<string, unknown> }).actions?.[action];
    assert(`An action named '${action}' was not found in ${target}`, Boolean(unchecked));
    assert(`An action named '${action}' was found in ${target}, but is not a function`, typeof unchecked === 'function');
    fn = value;
```

`typeof action` should already be string yeah? The guard on 398 should have narrowed it. Some unnecessary question mark/unchecked casts here.